Src/LiquidProjections/Dispatcher.cs

```diff
@@ -36,44 +39,98 @@ public Dispatcher(CreateSubscription createSubscription)
             {
                 LogProvider.GetLogger(typeof(Dispatcher)).FatalException(
                     "Projector exception was not handled. Event subscription has been cancelled.",
                     exception);
 
                 info.Subscription?.Dispose();
-            }
+            });
         }
 
         private async Task HandleUnknownCheckpoint(SubscriptionInfo info, Func<IReadOnlyList<Transaction>, SubscriptionInfo, Task> handler, SubscriptionOptions options)
         {
             if (options.RestartWhenAhead)
             {
-                try
-                {
-                    info.Subscription?.Dispose();
+                info.Subscription?.Dispose();
 
+                await ExecuteWithPolicy(info, async () =>
+                {
                     await options.BeforeRestarting();
 
                     Subscribe(null, handler, options);
-                }
-                catch (Exception exception)
+                }, abort: exception =>
                 {
                     LogProvider.GetLogger(typeof(Dispatcher)).FatalException(
                         "Failed to restart the projector.",
                         exception);
+                }, ignore: () => Subscribe(null, handler, options));
+            }
+        }
+
+        private async Task ExecuteWithPolicy(SubscriptionInfo info, Func<Task> action, Action<Exception> abort, Action ignore = null)
+        {
+            bool retry = true;
+            do
+            {
+                try
+                {
+                    await action();
+                    retry = false;
+                }
+                catch (Exception exception)
+                {
+                    ExceptionResolution resolution = await ExceptionHandler(exception, info);
+                    switch (resolution)
+                    {
+                        case ExceptionResolution.Abort:
+                            abort(exception);
+                            retry = false;
+                            break;
+
+                        case ExceptionResolution.Retry:
+                            break;
+
+                        case ExceptionResolution.Ignore:
+                            retry = false;
+                            ignore?.Invoke();
+                            break;
+                    }
                 }
             }
+            while (retry);
         }
     }
 
+    /// <summary>
+    /// Defines the signature for a method that handles an exception while dispatching transactions.
+    /// </summary>
+    /// <param name="exception">The exception that was caught by the <see cref="Dispatcher"/></param>
+    /// <param name="info">Information about the subscription.</param>
+    /// <returns>
+    /// Instructs the <see cref="Dispatcher"/> on how to resolve the exception.
+    /// </returns>
+    public delegate Task<ExceptionResolution> HandleException(Exception exception, SubscriptionInfo info);

```

Comment:
Should we maybe allow the handler to know some context about retrying progress? Like something allowing the handler to retry the same commit only 3 times, for example.